File: 142-Linked-List-Cycle-II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	slow := head
	fast := head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			res := head
			for res != slow {
				res = res.Next
				slow = slow.Next
			}
			return res
		}
	}

	return nil
}

File: 142-Linked-List-Cycle-II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        try:
            slow = head
            fast = head.next
            while slow != fast :
                slow = slow.next
                fast = fast.next.next

            s1 = head
            while s1 != slow.next:
                s1 = s1.next
                slow = slow.next
            return s1
        except:
            return None
